TIP-0002/README.md

```diff
@@ -0,0 +1,99 @@
+
+## Specification
+
+### Introduction
+
+In the Topl ecosystem, a lock address is a data structure that stores the evidence of a Quivr proposition. Besides the evidence, the lock address data structure stores the network and ledger identifier. This extra information is useful for improving usability and tooling. We bundle together the aforementioned information and then encode into a human readable string. The end user commonly perceives this string as “the address”.
+
+### Binary Format
+
+We encode the information in the address in an array of bytes in the following form:
+
+
+
+```
+  4 bytes       4 bytes      32 bytes
+<-----------> <---------> <-------------> 
+┌────────────┬───────────┬───────────────┐
+│ network id │ ledger id │ evidence      │
+└────────────┴───────────┴───────────────┘
+```
+
+
+
+We reserved the following values for the standard Topl Networks in the first 4 bytes as the network identifier.
+
+| **Network Name**      | **Byte Values** | **Address prefix** |
+| --------------------- | --------------- | ------------------ |
+| Main Network          | `0x8A11054C`    | `mtet`             |
+| Valhalla Test Network | `0xA5BF4108`    | `vtet`             |
+| Private Network       | `0x934B1900`    | `ptet`             |
+
+When encoded, the addresses will feature the prefix in the column labeled **Address prefix**.
+
+The next four bytes are the ledger identifier. We reserve the following value for the main ledger.
+
+| **Ledger Name** | **Byte Values** | **Remarks**                                                  |
+| --------------- | --------------- | ------------------------------------------------------------ |
+| Main Ledger     | `0xE7B07A00`    | This number has the nice property of being encoded to “main” when encoding in the main net and thus forcing the prefix to be `mtetmain` in main net addresses. However, it does not work that way for the other networks prefixes. |
+
+### Encoding
+
+We do the encoding of the data structure of the lock address using the Base58Check algorithm that is used for Bitcoin and other crypto currencies. The Base58Check algorithm takes as input the binary representation and returns an array of bytes.
+
+1. First compute the double sha256 hash of the payload. We have `doubleHash = sha256(sha256(payload))`.
+2. Take the first 4 bytes of the double sha256. We have `checksum = doubleHash.take(4)`.
+3. Concatenate the payload to the checksum and encode in [Base58](https://tools.ietf.org/html/draft-msporny-base58-02). We have `encoded = toBase58(payload.concat(checksum))`.
+
+### Examples
+
+The following are valid addresses for the different networks.
+
+#### Example 1
+
+Assuming the payload is a sequence of 32 bytes where all bytes are 0 (`payload = 0x0000000000000000000000000000000000000000000000000000000000000000`), the following are valid encoded addresses for Main, Valhalla and Private network respectively.
+
+- `mtetmain1y1Rqvj9PiHrsoF4VRHKscLPArgdWe44ogoiKoxwfevERNVgxLLh`

```

Comment:
There is a proposal to move Bech32, but Google agrees mostly that Base58 check is the one. Maybe Google is completely wrong also :/. There seems to be no consensus. What is your suggestion? Should we move to Bech32? Maybe do more research?

Reply:
🤕 thank you, I typed CBOR and meant Bech32. hmm... looking more into it, I see that Bitcoin and Cardano have both adopted Bech32 for including the human readable prefix after initially starting with Base58 so going that route seems more future-proof. However, since there are examples of moving from one to the other, what you have here is probably easier instead of introducing another new thing. 

Here are some of the links I was looking over to see what the standard approaches could be.
Solana - Base58 
https://docs.solana.com/api/websocket#accountsubscribe
https://solana-labs.github.io/solana-web3.js/classes/PublicKey.html

Tezos - Base58 
https://gitlab.com/tezos/tezos/-/issues/93

Scala Bech32 impl
https://gist.github.com/gafiatulin/cca9a375a127fb0a6e51efcfc0f83516

Cardano - Bech32
https://cips.cardano.org/cips/cip5/
https://cips.cardano.org/cips/cip19/
https://github.com/input-output-hk/bech32/blob/master/README.md

Bitcoin - Base58 + Bech32
https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki
https://github.com/sipa/bips/blob/bip-bech32m/bip-0350.mediawiki

In the case of Bitcoin they also introduced a new type of transaction when switching to Bech32. So P2PKH (pay-to-pubkey hash) became P2WPKH (pay-to-witness pubkey hash), the new type of transaction and the the encoding switch seem to be conflated quite a bit in what I can find through Google.

Reply:
A new task was created to analyze and include all of @scasplte2 remarks into the next iteration of the tip. The task is TSDK-486